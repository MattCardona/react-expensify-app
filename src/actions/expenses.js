import uuid from 'uuid';
import {database} from '../firebase/firebase.js';

// ADD_EXPENSE
const addExpense = (expense) => ({
  type: 'ADD_EXPENSE',
  expense
});

const startAddExpense = (expenseData = {}) => {
  return (dispatch) => {
    const {description = '', note = '', amount = 0, createdAt = 0} = expenseData;
    const expense = {description, note, amount, createdAt};

    return database.ref('expenses').push(expense).then((ref) => {
      dispatch(addExpense({
        id: ref.key,
        ...expense
      }));
    });
  };
};

// REMOVE_EXPENSE
const removeExpense = ({id} = {}) => ({
  type: 'REMOVE_EXPENSE',
  id
});

const startRemoveExpense = ({id} = {}) => {
  console.log(`Hey this id: id from actions ${id.id}`);
  return (dispatch) => {
    return database.ref(`expenses/${id}`).remove().then(() => {
      console.log('Data was removed');
      dispatch(removeExpense({id}));
    }).catch((e) => {
      console.log(`Their was a error: ${e}`);
    });
  }
};

// EDIT_EXPENSE
const editExpense = (id, updates) => ({
  type: 'EDIT_EXPENSE',
  id,
  updates
});


//SET_EXPENSES
const setExpenses = (expenses) => ({
  type: 'SET_EXPENSES',
  expenses
});

const startSetExpenses = () => {
  return (dispatch) => {
    return database.ref('expenses').once('value').then((snapshot) => {
      const dbExpenses = [];
      snapshot.forEach((child) => {
        dbExpenses.push({
          id: child.key,
          ...child.val()
        });
      dispatch(setExpenses(dbExpenses));
      });
    });

  }
};

export {addExpense, startAddExpense, removeExpense, startRemoveExpense, editExpense, setExpenses, startSetExpenses};