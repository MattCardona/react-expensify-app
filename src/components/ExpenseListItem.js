import React from 'react';
import {connect} from 'react-redux';
import {Link} from 'react-router-dom';

 // onClick={props.dispatch(removeExpense({id: id}))}
const ExpenseListItem = ({dispatch, id, description, amount, createdAt})=> (
  <div>
    <Link to={"/edit/" + id }><h3>{description}</h3></Link>
    <p>This expense is: {description}. The current balance is: {amount}. and it was created on: {createdAt}</p>
  </div>
);

const ConnectedExpenseListItem = connect()(ExpenseListItem);

export {ConnectedExpenseListItem, ExpenseListItem};