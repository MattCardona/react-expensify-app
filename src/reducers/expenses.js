const expensesReducerDefaultState = [];

const expensesReducer = (state = expensesReducerDefaultState, action) => {
  switch(action.type) {
    case 'ADD_EXPENSE':
      return [...state,action.expense]
    case 'REMOVE_EXPENSE':
      return state.reduce((memo, expense) => {
        if(expense.id !== action.id){
          memo.push(expense);
          return memo;
        }
        return memo;
      }, []);
    case 'EDIT_EXPENSE':
      return state.reduce((memo, expense) => {
        if(expense.id === action.id){
          memo.push({...expense, ...action.updates});
          return memo;
        }
        memo.push(expense);
        return memo;
      }, []);
    default:
      return state;
  }
};

export {expensesReducer};