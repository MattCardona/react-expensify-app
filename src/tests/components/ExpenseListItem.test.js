import React from 'react';
import {shallow} from 'enzyme';
import {ExpenseListItem} from '../../components/ExpenseListItem.js';
import expenses from '../fixtures/expenses.js';

test("should render ExpenseListItem correctly", () => {
  const wrapper = shallow(<ExpenseListItem key={expenses[1].id} {...expenses[1]} />);
  expect(wrapper).toMatchSnapshot();
});