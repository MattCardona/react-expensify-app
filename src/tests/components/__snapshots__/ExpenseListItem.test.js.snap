// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseListItem correctly 1`] = `
<div>
  <Link
    replace={false}
    to="/edit/2"
  >
    <h3>
      Rent
    </h3>
  </Link>
  <p>
    This expense is: 
    Rent
    . The current balance is: 
    109500
    . and it was created on: 
    -345600000
  </p>
</div>
`;
